### PR TITLE
Allow `skip_failed_rules` to skip  buggy logical plan rules that have a schema mismatch

### PR DESCRIPTION
**Pull request by smiklos**
## Which issue does this PR close?

In connection to #7241.

## Rationale for this change

If we allow buggy logical plans to be skipped then plans that don't fail but produce different schemas should also be skippable. 

## What changes are included in this PR?

Moved checking schema equality before handling logical plan errors so that the same option `skip_failed_rules` can be applied to these rules as well

## Are these changes tested?
Added a new test for this case. 

## Are there any user-facing changes?

The error message changed slightly. 

Don't think so

Nope

### CODE REVIEW

File: datafusion/optimizer/src/optimizer.rs [hunk 1]
```diff
@@ -288,11 +288,17 @@ impl Optimizer {
...
-
+                let result =
+                    self.optimize_recursively(rule, &new_plan, config)
+                        .and_then(|plan| {
+                            plan.map(|p| {
+                                assert_schema_is_the_same(rule.name(), &new_plan, &p)
+                                    .map(|_| Some(p))
+                            })
+                            .unwrap_or(Ok(None))
+                        });
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:299]: Here is another way of writing this logic which I think might be clearer:

```suggestion
                let result =
                    self.optimize_recursively(rule, &new_plan, config)
                    .and_then(|plan| {
                        if let Some(plan) = &plan {
                            assert_schema_is_the_same(rule.name(), &new_plan, &plan)?;
                        }
                        Ok(plan)
                    });
```

**Review by alamb — APPROVED** [review]
Thank you @smiklos  -- this is a nice improvement. I changed the title of this PR a little and have a suggestion on how to make the code cleaner, but it could also be done as a follow on OR

I also verified that the tests fail without this code change

```

failures:

---- optimizer::tests::generate_different_schema stdout ----
thread 'optimizer::tests::generate_different_schema' panicked at 'assertion failed: `(left == right)`
  left: `"Optimizer rule 'get table_scan rule' failed\ncaused by\nget table_scan rule\ncaused by\nInternal error: Failed due to generate a different schema, original schema: DFSchema { fields: [], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"a\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"b\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"c\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker"`,
 right: `"get table_scan rule\ncaused by\nInternal error: Failed due to generate a different schema, original schema: DFSchema { fields: [], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"a\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"b\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"c\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker"`', datafusion/optimizer/src/optimizer.rs:494:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
   2: core::panicking::assert_failed_inner
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:274:17
   3: core::panicking::assert_failed
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:228:5
   4: datafusion_optimizer::optimizer::tests::generate_different_schema
             at ./src/optimizer.rs:494:9
   5: datafusion_optimizer::optimizer::tests::generate_different_schema::{{closure}}
             at ./src/optimizer.rs:486:36
   6: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- optimizer::tests::skip_generate_different_schema stdout ----
thread 'optimizer::tests::skip_generate_different_schema' panicked at 'called `Result::unwrap()` on an `Err` value: Context("get table_scan rule", Internal("Failed due to generate a different schema, original schema: DFSchema { fields: [], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"a\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"b\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: Some(Bare { table: \"test\" }), field: Field { name: \"c\", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }"))', datafusion/optimizer/src/optimizer.rs:517:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
   2: core::result::unwrap_failed
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/result.rs:1651:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/result.rs:1076:23
   4: datafusion_optimizer::optimizer::tests::skip_generate_different_schema
             at ./src/optimizer.rs:517:9
   5: datafusion_optimizer::optimizer::tests::skip_generate_different_schema::{{closure}}
             at ./src/optimizer.rs:510:41
   6: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    optimizer::tests::generate_different_schema
    optimizer::tests::skip_generate_different_schema
```

File: datafusion/optimizer/src/optimizer.rs (hunk 1 above)
Comment (smiklos) [reply datafusion/optimizer/src/optimizer.rs]: Thanks @alamb , I've added your suggestion. Defy looks much better this way 🥇

**Comment by alamb** [comment]
I took the liberty of pushing a fix for clippy to this branch as well

**Pull request merged**